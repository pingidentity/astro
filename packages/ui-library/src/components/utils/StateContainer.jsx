import React from "react";
import PropTypes from "prop-types";
import Utils from "../../util/Utils";

const mapNothing = () => ({});

const defaultGetPassedProps = props => props;
const containerGetPassedProps = props => props.passedProps;

export const createProgressiveState = stateDefs => ({
    passedProps,
    initialState,
    boundSetState,
    getPassedProps = defaultGetPassedProps,
}) => {
    const isControlled = name => passedProps[name] === undefined;

    // figure out which props will be controlled and create initial state for those
    const pState = stateDefs.reduce((state, { name, initial }) => {
        if (isControlled(name)) {
            return {
                ...state,
                [name]: initialState[name] === undefined ? initial : initialState[name],
                // the initialState prop should override the defined initial value
                // but only if this is a controlled prop
            };
        }
        return state;
    }, {});

    // make sure the keys provided in initialState are actually being controlled
    if (!Utils.isProduction()) {
        Object.keys(initialState).forEach(key => {
            if (pState[key] === undefined) {
                console.warn(
                    `Initial state provided for prop "${key}" that is not being controlled by the component`
                );
            }
        });
    }

    // generate custom callbacks with transform functions
    const makeTransformCallbacks = (name, callbacks = []) =>
        callbacks.reduce(
            (transformCallbacks, { name: cbName, transform, passTransformedValue = false, }) => ({
                ...transformCallbacks,
                [cbName]: (value, e) => {
                    // set the state with the transformed value
                    boundSetState((state, props) => {
                        const transformedValue = transform(value, state[name], getPassedProps(props));

                        if (passedProps[cbName]) {
                            passedProps[cbName](passTransformedValue ? transformedValue : value, e);
                        }

                        return { [name]: transformedValue };
                    });
                }
            }), {}
        );

    // get the callbacks for a single state definition
    const makeCallbacksFromDef = ({ name, setter, callbacks }) => {
        // add any custom callbacks
        const customTransformCallbacks = makeTransformCallbacks(name, callbacks);

        // return with setter if defined
        return setter ? {
            [setter]: (value, e) => {
                boundSetState({ [name]: value });
                // if a callback is provided, we'll still execute it
                if (passedProps[setter]) {
                    passedProps[setter](value, e);
                }
            },
            ...customTransformCallbacks
        } : customTransformCallbacks;
    };

    // create all the callbacks
    const pCallbacks = stateDefs.reduce((returnCallbacks, def) => {
        if (!isControlled(def.name)) {
            return returnCallbacks;
        }
        return { ...returnCallbacks, ...makeCallbacksFromDef(def) };
    }, {});

    return [pState, pCallbacks];
};

class StateContainer extends React.Component {
    static propTypes = {
        stateDefs: PropTypes.arrayOf(
            PropTypes.shape({
                name: PropTypes.string.isRequired, // name of the props
                initial: PropTypes.any.isRequired, // starting value
                setter: PropTypes.string, // name of the callback in the target component
                callbacks: PropTypes.arrayOf( // when you need something different than a simple setter
                    PropTypes.shape({
                        name: PropTypes.string,
                        // Decides whether to pass the transformed value to the callback or just the value that
                        // the base component passes.
                        passTransformedValue: PropTypes.bool,
                        transform: PropTypes.func, // this function will be applied to the value and current state
                    })
                ),
            })
        ),
        initialState: PropTypes.object,
        passedProps: PropTypes.object,
        mapToProps: PropTypes.func,
    };

    static defaultProps = {
        stateDefs: [],
        initialState: {},
        passedProps: {},
        mapToProps: mapNothing,
    };

    constructor({ stateDefs, initialState, passedProps }) {
        // the constructor has two main goals
        // 1) populate the state for any props that will be controlled
        // 2) create callback functions for those props
        super();

        const [ state, callbacks ] = createProgressiveState(stateDefs)({
            initialState, passedProps, boundSetState: this.setState.bind(this), getPassedProps: containerGetPassedProps
        });

        this.state = state;
        this.callbacks = callbacks;
    }

    render() {
        const { children, mapToProps, passedProps } = this.props;

        // alert dev if there are state/prop conflicts
        if (!Utils.isProduction()) {
            Object.keys(this.state).forEach(key => {
                if (passedProps[key] !== undefined) {
                    console.warn(
                        `The prop "${key}" is being controlled by the component because it was not initially set. ` +
                        `The newly-received value will be ignored.`
                    );
                }
            });
        }

        return children({ ...passedProps, ...this.callbacks, ...this.state, ...mapToProps(this.state, passedProps) });
    }
}

export const inStateContainer = (stateDefs, mapToProps) => WrappedComponent => ({ initialState, ...props }) => (
    <StateContainer stateDefs={stateDefs} mapToProps={mapToProps} initialState={initialState} passedProps={props}>
        {containerProps => <WrappedComponent {...containerProps} />}
    </StateContainer>
);

export const toggleTransform = (value, current) => (!current);

export default StateContainer;