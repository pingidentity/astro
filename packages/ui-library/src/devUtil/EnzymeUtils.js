import { mount } from "enzyme";


export const snapshotDataIds = node => {
    const elements = node.find("[data-id]").hostNodes(); // Find only actual HTML elements

    const ids = elements.map(element => element.prop("data-id")).filter(item => item !== "");
    expect(ids).toMatchSnapshot();
};

export const mountSnapshotDataIds = node => snapshotDataIds(mount(node));
