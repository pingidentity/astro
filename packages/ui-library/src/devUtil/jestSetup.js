
// This is needed for testing the modal and DetailsTooltip component.
global.getSelection = jest.fn();
global.getSelection.mockReturnValue({
    toString: () => "",
});
global.document.createRange = () => ({
    setStart: () => {},
    setEnd: () => {},
    commonAncestorContainer: {
        nodeName: "BODY",
        ownerDocument: document,
    },
});
