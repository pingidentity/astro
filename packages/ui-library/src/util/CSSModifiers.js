export const lightInputs = "modifier_light-inputs";
export const darkInputs = "modifier_dark-inputs";
export const stripInputMargins = "modifier_strip-input-margins";

export default { lightInputs, darkInputs, stripInputMargins };
