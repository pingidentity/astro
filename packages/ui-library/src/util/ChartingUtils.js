import Color from "color";

const splitData = ({ id, data }) => data.map(point => ({ [id]: point }));

/**
 * Takes data from the format of [{ id: [data] }, { id2: [data2] }] and converts
 * it to the format of [[{ id: dataPoint }, { id2: dataPoint to }]]
 * All incoming arrays of data must be of the same length
 */
export const toRechartsDataFormat = dataSet => dataSet.reduce((acc, data) => {
    const split = splitData(data);
    return acc.length > 0
        ? acc.map((frame, index) => ({ ...frame, ...split[index] }))
        : split
    ;
}, []);

const addToExisting = (index, data, acc) => {
    const { id, data: currentData } = acc[index];
    return {
        data: [
            ...currentData,
            data,
        ],
        id
    };
};

/**
 * Takes data from the format of [[{ id: dataPoint }, { id2: dataPoint to }]] and converts
 * it to the format of [{ id: [data] }, { id2: [data2] }]
 */
export const fromRechartsDataFormat = dataSet => dataSet.reduce((acc, frame) =>
    Object.entries(frame).reduce((frameAcc, [id, data]) => {
        const idIndex = frameAcc.findIndex(entry => entry.id === id);
        return idIndex > -1
            ? [
                ...idIndex > 0 ? frameAcc.slice(0, idIndex) : [],
                addToExisting(idIndex, data, frameAcc),
                ...idIndex < frameAcc.length ? frameAcc.slice(idIndex + 1, frameAcc.length) : []
            ]
            : [...frameAcc, { id, data: [data] }];
    }, acc)
, []);

/**
 * Generates color props for chartes based on a given
 * "seed" color provided by the user
 */
export const generateTheme = (seedColor, data) => {
    const step = 360 / data.length;

    const dataColors = data.map((item, index) => {
        return {
            ...item,
            color: Color(seedColor).rotate(step * index).hex(),
        };
    });

    const contrastedColor = Color(seedColor).rotate(180).hex();

    return {
        highlightColor: contrastedColor,
        referenceLineColor: contrastedColor,
        referenceLabelColor: "#676D74",
        dataColors: dataColors,
    };
};

/**
 * Returns coordinates for drawing evenly spaced lines with a min and max offset
 */

export const getEvenLineCoords = (height, numLines = 5) => {
    //offsets are set to what recharts uses by default
    const minOffset = 20;
    const maxOffset = 5;

    const actualMin = 0 + minOffset;
    const actualMax = height - maxOffset;

    const range = actualMax - actualMin;

    const lines = Array.from( { length: numLines }, (v, i) => {
        return ( range / (numLines - 1) * (i) + minOffset);
    });

    return lines;
};
