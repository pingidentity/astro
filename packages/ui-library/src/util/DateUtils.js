/*eslint-disable valid-jsdoc*/

"use strict";
const moment = require("moment");
const dateFormat = "YYYY-MM-DD";
const timeFormat = "hh:MM:SSa";

const DateUtils = {
    /**
     * @alias module:util/Utils.formatDate
     * @desc Format the given timestamp in milliseconds as a YYYY-MM-DD date.
     *
     * @param {number} timestamp
     *     The timestamp (UNIX time * 1000)
     * @returns {string}
     *     Formatted date
     */
    formatDate: function (timestamp) {
        return moment.utc(timestamp).format(dateFormat);
    },

    /**
     * @alias module:util/Utils.formatTime
     * @desc Format the given timestamp in milliseconds as a YYYY-MM-DD hh:MM:SSa date.
     *
     * @param {number} timestamp
     *     The timestamp (UNIX time * 1000)
     * @returns {string}
     *     Formatted date+time
     */
    formatDateTime: function (timestamp) {
        return moment.utc(timestamp).format(`${dateFormat} ${timeFormat}`);
    },

    /**
     * @alias module:util/Utils.formatTime
     * @desc Format the given timestamp in milliseconds as a hh:MM:SSa time.
     *
     * @param {number} timestamp
     *     The timestamp (UNIX time * 1000)
     * @returns {string}
     *     Formatted time
     */
    formatTime: function (timestamp) {
        return moment.utc(timestamp).format(timeFormat);
    },
};

module.exports = DateUtils;