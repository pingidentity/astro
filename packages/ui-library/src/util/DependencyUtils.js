import React from "react";

/**
 * @module util/DependencyUtils
 * @desc Functions for showing nice errors when loading a component that has a requirement above our standard dependencies.
 */

/**
 * @alias module:util/DependencyUtils.usesStableContext
 *
 * @desc Throws an error if the React 16.3+ context API isn't available.
 *
 * @example
 *   usesStableContext()
 */

export const usesStableContext = (component) => {
    if (!React.createContext) {
        console.error(`${component} requires React 16.3`);
    }
};

/**
 * @alias module:util/DependencyUtils.usesFragments
 *
 * @desc Throws an error if React fragments aren't available.
 *
 * @example
 *   usesStableContext()
 */

export const usesFragments = (component) => {
    if (!React.Fragment) {
        console.error(`${component} requires React 16`);
    }
};
