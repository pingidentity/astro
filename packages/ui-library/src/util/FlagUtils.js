import PropTypes from "prop-types";

export const allFlags = [
    "pass-value",
    // v4 flags
    "p-stateful",
    "use-portal",
    "add-button-margin",
    "expandable-row-class",
    "fix-discard-button",
    "fixed-messages-constants",
    "true-default",
    "v4",
    "new-alert-modal",
];

export const flagsPropType = PropTypes.arrayOf(PropTypes.oneOf(allFlags));

export const getFlags = ({ props, context }) => props.flags || (context && context.flags) || [];

export const hasFlag = (component, flag) => {
    const flags = getFlags(component);
    return flags.includes(flag);
};
