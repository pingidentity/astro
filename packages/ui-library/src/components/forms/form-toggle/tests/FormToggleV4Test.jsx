jest.dontMock("../v2");

import { allFlags } from "../../../../util/FlagUtils";
import StateContainer from "../../../utils/StateContainer";
import { mountSnapshotDataIds } from "../../../../devUtil/EnzymeUtils";

describe("Toggle v4", function () {
    const React = require("react");
    const ReactTestUtils = require("react-dom/test-utils");
    const _ = require("underscore");
    const TestUtils = require("../../../../testutil/TestUtils");
    const Toggle = require("../v2");

    function getComponent (p) {
        const props = _.defaults(p || {}, {
            stateless: true,
            onToggle: jest.fn(),
            flags: allFlags,
        });

        return TestUtils.renderInWrapper(<Toggle {...props} />);
    }

    it("data-id's don't change", () => {
        mountSnapshotDataIds(<Toggle flags={allFlags} />);
    });

    it("renders the component with default data-id", function () {
        const component = getComponent();

        expect(TestUtils.findRenderedDOMNodeWithDataId(component, "toggle")).toBeTruthy();
    });

    it("stateless: renders additional CSS classes from className prop", function () {
        const component = getComponent({ className: "someClass" }),
            toggle = TestUtils.findRenderedDOMNodeWithDataId(component, "toggle"),
            toggleCSS = toggle.className.split(" ");

        expect(_.contains(toggleCSS, "someClass")).toBe(true);
    });

    it("stateless: defaults to toggled false", function () {
        const component = getComponent();
        const toggle = TestUtils.findRenderedDOMNodeWithDataId(component, "toggle");
        const toggleCSS = toggle.className.split(" ");

        expect(_.contains(toggleCSS, "selected")).toBe(false);
    });

    it("stateful: defaluts to toggled false", function () {
        const component = getComponent({ stateless: false });
        const toggle = TestUtils.findRenderedDOMNodeWithDataId(component, "toggle");
        const toggleCSS = toggle.className.split(" ");

        expect(_.contains(toggleCSS, "selected")).toBe(false);
    });

    it("stateless: accepts default toggled state", function () {
        const component = getComponent({ toggled: true });
        const toggle = TestUtils.findRenderedDOMNodeWithDataId(component, "toggle");
        const toggleCSS = toggle.className.split(" ");

        expect(_.contains(toggleCSS, "selected")).toBe(true);
    });

    it("stateful: accepts default toggled state", function () {
        const component = getComponent({
            stateless: false,
            toggled: true
        });
        const toggle = TestUtils.findRenderedDOMNodeWithDataId(component, "toggle");
        const toggleCSS = toggle.className.split(" ");

        expect(_.contains(toggleCSS, "selected")).toBe(true);
    });

    it("stateless: triggers onToggle callback when clicked", function () {
        const component = getComponent();
        const toggle = TestUtils.findRenderedDOMNodeWithDataId(component, "toggle");

        ReactTestUtils.Simulate.click(toggle);

        expect(component.props.children.props.onToggle).toBeCalled();
    });

    it("does not error with default onToggle function", function () {
        const component = TestUtils.renderInWrapper(<Toggle />);
        const toggle = TestUtils.findRenderedDOMNodeWithDataId(component, "toggle");

        ReactTestUtils.Simulate.click(toggle);

        expect(component).toBeTruthy();
    });

    it("stateful: _handleToggle callback changes toggled state when clicked", function () {
        const component = getComponent({ stateless: false });
        const componentRef = ReactTestUtils.findRenderedComponentWithType(component, StateContainer);
        const toggle = TestUtils.findRenderedDOMNodeWithDataId(component, "toggle");

        expect(componentRef.state.toggled).toBe(false);

        ReactTestUtils.Simulate.click(toggle);

        expect(componentRef.state.toggled).toBe(true);
    });

    it("stateless: does not trigger onToggle callback when disabled", function () {
        const component = getComponent({ disabled: true, "data-id": "toggle" });
        const toggle = TestUtils.findRenderedDOMNodeWithDataId(component, "toggle");

        ReactTestUtils.Simulate.click(toggle);

        expect(component.props.children.props.onToggle).not.toBeCalled();
    });

    it("renders locked toggle", function() {
        const component = getComponent({ status: Toggle.Status.LOCKED });

        const locked = TestUtils.findRenderedDOMNodeWithClass(component, "locked");
        expect(locked).toBeTruthy();
    });

});
