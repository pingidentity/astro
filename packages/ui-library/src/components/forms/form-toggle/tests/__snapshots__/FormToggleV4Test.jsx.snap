// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toggle v4 data-id's don't change 1`] = `
Array [
  "toggle",
]
`;
